Update & optimize log scrolling functionality to go with theme-f1

benjaminapetersen opened the pull request:
@sg00dwin @spadgett 
- directive, when rendered, will grab the scrollable node & cache
- changes to fns in the logViewer controller to indicate scoll DOM node vs window
- optimizations to logLinks fns, cache selectors & use raw DOM when possible
- added id to DOM to allow for document.getElementById, though since this node is now cached once it may not be as necessary to use
- adds BREAKPOINTS constant to app.js so we have one place to update if the CSS breakpoints ever change

the logLinks.scrollBottom() function was tricky & ended up still needing jQuery to normalize the doc.height & win.height to get it to work properly.  Both are selected once and cached, so hopefully running the function is as minimal on performance as possible.

I'll take suggestions!

sg00dwin committed:
Rework console theme

Changes sidebar/navbar colors

markup and styles to support new structure

post switch back to use current sidebar nav
mostly there... needs sidebar-left width adjustments and mobile styling

Top header and sidebar/nav are styled and functional at all responsive widths
- mobile nav uses current dropdown toggle

added header to all pages, customize scrollbars, project-menu converted dark

fix follow log scroll top/bottom links

Make default-header pages (create,projects,...) have inner scroll by adding div.middle
also add arrow to dropdown-menu for utility links

Project pages now have home icon in place of logo

Mobile nav is amalgamated into a single bar

top and side bar navigation for both mobile and desktop

Add individual scrollable/non-scrollable sections to middle and right columns

all pages updated with substructure markup for scrolling

benjaminapetersen committed:
Update & optimize log scrolling functionality

benjaminapetersen commented:
Assuming Steve will want to pull this into his PR and we merge it all into Origin as one commit....

benjaminapetersen commented:
I'll have to rebase on Steve's latest commits I see, & build-assets....

spadgett commented on assets/app/scripts/services/logLinks.js:
```diff
@@ -5,34 +5,39 @@ angular.module('openshiftConsole')
+      var $doc = $(document);
+      var $win = $(window);
+
+      var scrollBottom = function(node, isWindow) {
```
Rather than passing `isWindow`, I'd rather just pass `null` for node.

```
var scrollBottom = function(node) {
    if(!node) {
        $window.scrollTo(null, $doc.height() - $win.height());
    } else {
        node.scrollTop = node.scrollHeight;
    }
}
```

benjaminapetersen replied:
I was trying to keep direct ref to window/document out of the service in case we ever write tests, so we can mock if needed.

spadgett replied:
I think we should cross that bridge when we come to it.

benjaminapetersen replied:
Not sure I like knowingly reducing testability, but I'll update!

spadgett replied:
In that case, I'd almost rather have two methods.

```
var scrollBottomNode = function(node) { ... }

var scrollBottomWindow = function(window) { ... }
```

Controller already has to know about this difference and handle the breakpoint, but it keeps the service methods dead simple.

spadgett commented on assets/app/views/directives/header/_navbar-utility.html:
```diff
@@ -1,4 +1,7 @@
 <ul class="nav navbar-nav navbar-right navbar-iconic">
+  <li>
+    <span id="outage" class="status-issue nav-item-iconic"></span>
+  </li>
```
Should this be here?

benjaminapetersen replied:
Rebasing & pushing shortly, think its difference of old code

sg00dwin replied:
same as above

spadgett commented on assets/app/scripts/services/logLinks.js:
```diff
@@ -5,34 +5,39 @@ angular.module('openshiftConsole')
+        if(isWindow) {
+          // ideally we would go native, but is fussy
+          // node.scrollTo(null, document.body.clientHeight - node.innerHeight);
+          $window.scrollTo(null, $doc.height() - $win.height());
```
Did you test this?

```
window.scrollTo(0, document.body.scrollHeight - document.body.clientHeight);
```

benjaminapetersen replied:
updated!

spadgett commented on assets/app/styles/_navbar-alt.less:
```diff
@@ -113,12 +113,21 @@
+      font-size: 17px;
+      -webkit-font-smoothing: antialiased;
+      -moz-osx-font-smoothing: grayscale;
+    }
```
More status-issue stuff...

benjaminapetersen replied:
same, rebasing and will see if this disapears

sg00dwin replied:
yes please rebase and these shouldn't be there anymore

spadgett commented on assets/app/scripts/directives/logViewer.js:
```diff
@@ -268,8 +278,42 @@ angular.module('openshiftConsole')
+                setScrollable();
+              };
+              // maintain the correct scrollable node
+              $(window).on('resize', setScrollable);
```
We already have a resize listener above. Can we hook into that?

benjaminapetersen replied:
We can, I was attempting to keep the behavior a bit more separated, there is a lot going on in this directive.  I can push it up into the single listener though.

spadgett replied:
Resize generates a lot of events. We should at least debounce.

benjaminapetersen replied:
I would probably prefer that approach!

benjaminapetersen commented:
Updated again:
- all DOM node selection is done once & cached for life of directive w/o jQuery
- removed the extra less & template stuff per @spadgett 
- normalized some of the event syntax (`.on('resize', onResize);` style always)
- reduced a `ng-include` which allowed further selector optimization
- debounced a resize listener

benjaminapetersen commented:
I believe we should be good to merge this into @sg00dwin branch unless either of you have any more thoughts!

@spadgett @sg00dwin

spadgett commented on assets/app/scripts/directives/logViewer.js:
```diff
@@ -69,46 +84,39 @@ angular.module('openshiftConsole')
-            };
 
-            var buffer = document.createDocumentFragment();
+            var buffer = clearBuffer();
```
The function is not really clearing a buffer. It's creating a doc fragment. I feel like pulling this one-liner into a function is making it harder to follow.

benjaminapetersen replied:
I saw a comment `// reset buffer` that was doing the same thing, so figured I'd use a named function to eliminate the comment.  I can return it, or perhaps use `resetBuffer()`?

spadgett replied:
Right, but it's called in other places as well where the reset buffer comment doesn't make sense. I think I preferred how it was before.

benjaminapetersen replied:
I like the abstraction of 'what i'm doing' over 'how i am doing it', but I'll roll it back!

spadgett commented on assets/app/scripts/directives/logViewer.js:
```diff
@@ -268,8 +286,45 @@ angular.module('openshiftConsole')
+
+
+              // scrollable node is window if mobile, else a particular DOM node.
+              var setScrollable = function() {
```
`setScrollable` sounds like you're enabling scrolling. `findScrollableNode`?

benjaminapetersen replied:
Yeah, i guess the link fn is doing the finding with the `document.getElementById` & `cacheLogNode`.  So `set`, `toggle`, `find`... not sure what the best word here is?

benjaminapetersen replied:
`detectScrollableNode()` ?

benjaminapetersen commented:
Rebased / cherry-picked on a new branch, will reopen this again in a minute.
